node/src/components/block_synchronizer/signature_acquisition.rs

```diff
@@ -109,3 +109,317 @@ impl SignatureAcquisition {
+            Acceptance::NeededIt
+        );
+        assert_equal(
+            signature_acquisition.have_signatures(),
+            [public_0, public_2, public_3],
+        );
+        assert_equal(
+            signature_acquisition.not_vacant(),
+            [public_0, public_2, public_3],
+        );
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().map(|(p, _)| p),
+        );
+        // The total signed weight is 8/10, which is higher than 2/3:
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Strict
+        );
+    }
+
+    #[test]
+    fn register_pending_has_the_expected_behavior() {
+        let rng = &mut TestRng::new();
+        let validators = validators(rng, 4);
+        let era_id = EraId::new(rng.gen());
+        let block_hash = BlockHash::random(rng);
+        let weights = EraValidatorWeights::new(
+            era_id,
+            validators
+                .iter()
+                .enumerate()
+                .map(|(i, (public, _))| (public.clone(), (i + 1).into()))
+                .collect(),
+            Ratio::new(1, 10), // Low finality threshold
+        );
+        assert_eq!(U512::from(10), weights.get_total_weight());
+        let mut signature_acquisition =
+            SignatureAcquisition::new(validators.iter().map(|(p, _)| p.clone()).collect());
+
+        // Set the validator #0 weighting 1 as pending:
+        let (public_0, secret_0) = validators.get(0).unwrap();
+        signature_acquisition.register_pending(public_0.clone());
+        assert_equal(signature_acquisition.have_signatures(), []);
+        assert_equal(signature_acquisition.not_vacant(), [public_0]);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().skip(1).map(|(p, _s)| p).collect_vec(),
+        );
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Insufficient
+        );
+
+        // Sign it:
+        let finality_signature =
+            FinalitySignature::create(block_hash, era_id, secret_0, public_0.clone());
+        assert_matches!(
+            signature_acquisition.apply_signature(finality_signature, &weights),
+            Acceptance::NeededIt
+        );
+        assert_equal(signature_acquisition.have_signatures(), [public_0]);
+        assert_equal(signature_acquisition.not_vacant(), [public_0]);
+        assert!(signature_acquisition.have_no_vacant() == false);
+        assert_equal(
+            signature_acquisition.not_pending(),
+            validators.iter().map(|(p, _)| p),
+        );
+        assert_eq!(
+            signature_acquisition.signature_weight(),
+            SignatureWeight::Weak
+        );
+    }
+
+    /*
```

Comment:
BTW, it's kinda bad that we can add wrong signatures (for example a mismatch of public/secret keys) and everything works fine. If someone forgets to check the signature before, it can lead to severe vulnerabilities. I would have encoded this in the type system to be sure nothing was forgotten.

Reply:
Yes, I agree. We can consider such refactor in the future.